R/linear_pool_quantile.R

```diff
@@ -46,10 +41,21 @@ linear_pool_quantile <- function(model_out_tbl, weights = NULL,
       .groups = "drop"
     ) |>
     tidyr::unnest("pred_qs") |>
+    dplyr::group_split(dplyr::across(dplyr::all_of(weight_by_cols))) |>
+    purrr::map(.f = function(outputs_by_weight) {
+      if (!is.null(weights)) {
+        weight_temp <- outputs_by_weight$weight[1]
+        sample_index <- sample(1:(n_samples - 1), weight_temp * n_samples * nrow(weights), replace = TRUE)
+        outputs_by_weight[sample_index, ]
+      } else {
+        outputs_by_weight
+      }
+    }) |>
+    purrr::list_rbind() |>
     dplyr::group_by(dplyr::across(dplyr::all_of(task_id_cols))) |>
     dplyr::summarize(
       output_type_id = list(quantile_levels),
-      value = list(do.call("wtd.quantile", args = agg_args)),
+      value = list(do.call("quantile", args = agg_args)),
```

Comment:
It would be preferable to avoid the extra level of sampling in this operation. Additionally, a more minor point: I'm not sure why `nrow(weights)` shows up in the sampling step on line 48.

My suggestion is to write our own `weighted_quantile` function that accepts weights and performs an exact weighted quantile computation. Here is an outline of what that function could look like:

```
weighted_quantile <- function(x, weights, normalize, probs) {
  # check that x is numeric

  # check that weights is numeric

  # check that x and weights have the same length

  # if normalize = TRUE, normalize the weights
  if (normalize) weights <- weights / sum(weights)

  # check that the sum of weights is 1

  # check that probs is numeric, and all are > 0 and < 1

  # sort both x and weights in increasing order of x
  x_order <- order(x)
  x_sorted <- x[x_order]
  weights_sorted <- weights[x_order]
  
  # cumulative weights
  cum_weights <- cumsum(weights_sorted)

  # quantiles at each probability level:
  # the smallest x with cumulative weight at least as large as the desired probability level
  quantiles <- lapply(
    probs,
    function(prob) {
      x_sorted[min(which(cum_weights >= prob))]
    }
  ) |>
    unlist()

  return(quantiles)
}
```

Here's an example of calling this function, which could be the foundation of a unit test:

```
x <- c(9, 4.2, 3, 5, 6, 14, 1)
weights <- seq_along(x)

weighted_quantile(x, weights, normalize=TRUE, probs = seq(from = 0.1, to = 0.9, by = 0.1))
```

Reply:
To that minor point, i had added `nrow(weights)` to increase the number of samples per group so that we want to run into a situation where some group had very few samples. But I have since scrapped this version and implemented your `weighted_quantile` function instead